Fixed ida plugin working in IDA 9.x

ZERO-A-ONE opened the pull request:
Although I have already found the following work: https://github.com/NationalSecurityAgency/ghidra/issues/7392, which has attempted to fix idaxml's functionality in IDA 9.x, I still encountered runtime errors on my own IDA 9.0.

## 🎯 Overview

This pull request resolves critical compatibility issues preventing IDAPro XML plugins from working correctly in IDA 9.0 with Python 3. The main fixes address broken exception handling that was masking real errors and prepare the codebase for IDA 9.0 API changes.

## 🐛 Problems Solved

### 1. Broken Exception Display in Python 3

**Issue:** Exception information was showing as lambda function memory addresses instead of actual error messages, making debugging impossible.

**Example of broken output:**
```
***** Exception occurred: XML Exporter failed! *****
 <function <lambda> at 0x15256c680> <function <lambda> at 0x152469b20>
```

**Root cause:** Legacy Python 2/3 compatibility code created lambda functions to simulate Python 2's `sys.exc_type` and `sys.exc_value`, but the print statements weren't calling these functions.

### 2. IDA 9.0 API Breaking Changes

**Issue:** `TypeError: tinfo_t.__init__() got an unexpected keyword argument 'tid'`

**Root cause:** IDA 9.0 removed support for the `tid` parameter in `tinfo_t` constructor, requiring a two-step initialization pattern.

### 3. Obsolete Compatibility Code

**Issue:** Python 2/3 compatibility shims causing confusion since IDA 9.x only supports Python 3.

## ✅ Solutions Implemented

### Exception Handling Modernization

**Before:**
```python
# Broken compatibility layer
if sys.version_info.major >= 3:
    from idaxml import _exc_info
    sys.exc_value = lambda: _exc_info()[1]  # Returns function, not value!
    sys.exc_type = lambda: _exc_info()[0]   # Returns function, not value!

# Broken usage
except:
    print("\n" + msg + "\n", sys.exc_type, sys.exc_value)  # Prints functions
```

**After:**
```python
# Modern Python 3 approach
except Exception as e:
    print(f"\n{msg}\n{type(e).__name__}: {e}")  # Prints actual error
```

### API Migration Documentation

**IDA 8.x (deprecated):**
```python
tif = ida_typeinf.tinfo_t(tid=some_tid)  # No longer works
```

**IDA 9.0 (correct):**
```python
tif = ida_typeinf.tinfo_t()
if not tif.get_type_by_tid(some_tid):
    # Handle type loading failure
    return None
```

## 📁 Files Modified

| File | Changes |
|------|---------|
| `plugins/xml_exporter.py` | ✅ Fixed exception handling<br>✅ Removed compatibility code |
| `plugins/xml_importer.py` | ✅ Fixed exception handling<br>✅ Removed compatibility code |
| `loaders/xml_loader.py` | ✅ Fixed exception handling<br>✅ Removed compatibility code |
| `python/idaxml.py` | ✅ Fixed exception handling in 2 locations |

## 🧪 Testing Results

```bash
---------------------------------------------------------------------------------------------
Python 3.11.13 (main, Jun  5 2025, 08:25:22) [Clang 14.0.6 ] 
IDAPython 64-bit v9.0.0 final (serial 0) (c) The IDAPython Team <idapython@googlegroups.com>
---------------------------------------------------------------------------------------------

XML Exporter v5.0.2 : SDK 900 : Python : Aug 19 2025 23:09:48

-----------------------------------------------------------
Exporting XML <PROGRAM> document ....
Processing PROGRAM                 CPU time: 0.0537
Processing DATATYPES               CPU time: 0.7057
Processing MEMORY_MAP              CPU time: 3.5783
Processing REGISTER_VALUES         CPU time: 0.0006
Processing CODE                    CPU time: 0.2650
Processing DATA                    CPU time: 0.3727
Processing COMMENTS                CPU time: 0.9637
Processing PROGRAM_ENTRY_POINTS    CPU time: 0.0001
Processing SYMBOL_TABLE            CPU time: 1.0039
Processing FUNCTIONS               CPU time: 1.4628
Processing MARKUP                  CPU time: 3.7575
                             Total CPU time: 12.4148
--------------------------------------
PROGRAM                           1
INFO_SOURCE                       1
PROCESSOR                         1
COMPILER                          1
DATATYPES                         1
STRUCTURE                         5
MEMBER                         8178
REPEATABLE_CMT                 8178
UNION                            29
MEMORY_MAP                        1
MEMORY_SECTION                    7
MEMORY_CONTENTS                   4
REGISTER_VALUES                   1
REGISTER_VALUE_RANGE             14
CODE                              1
CODE_BLOCK                        2
DATA                              1
DEFINED_DATA                    112
TYPEINFO_CMT                   6642
COMMENTS                          1
COMMENT                          99
PROGRAM_ENTRY_POINTS              1
PROGRAM_ENTRY_POINT               5
SYMBOL_TABLE                      1
SYMBOL                         7309
FUNCTIONS                         1
FUNCTION                       6631
ADDRESS_RANGE                  7381
STACK_FRAME                    5413
MARKUP                            1
MEMORY_REFERENCE                 64
--------------------------------------
Total XML Elements:           50087
Database exported to: /Users/syc/Person/TMP/TmpAndroidProject/platform_channel/build/app/outputs/flutter-apk/app-release/lib/arm64-v8a/IDA.xml
```

Der-Systemfehler commented:
I might be able to test this in a few hours. I ran into this problem just yesterday so stumbled upon this fix. 
The last PR didn't work ( I assume because of the missing files) but I couldn't test this one yet.

ZERO-A-ONE commented:
> 3.10.5 is the Version I have - should I use a different one?
> 
> IDA Version is 9.1

Thank you for your testing. I have updated xml_exporter.py. I am using IDA 9.0 and Python 3.11. Regarding the TypeError: %X format: an integer is required, not NoneType situation, further confirmation is needed.

ZERO-A-ONE commented:
> Thank you for fixing! The fix helped with proper error handling, the nonetype seems to remain:
> 
> ```
> 
> Python 3.10.5 (tags/v3.10.5:f377153, Jun  6 2022, 16:14:13) [MSC v.1929 64 bit (AMD64)] 
> IDAPython 64-bit v9.1.0 (c) The IDAPython Team <idapython@googlegroups.com>
> -----------------------------------------------------------------------------------------
> 
> XML Exporter v5.0.2 : SDK 910 : Python : Aug 20 2025 15:22:57
> 
> -----------------------------------------------------------
> Exporting XML <PROGRAM> document ....
> Processing PROGRAM                 CPU time: 0.0886
> Processing DATATYPES               CPU time: 0.3353
> Processing MEMORY_MAP              CPU time: 116.4273
> Processing REGISTER_VALUES         CPU time: 0.0912
> Processing CODE                    CPU time: 12.7872
> Processing DATA                    CPU time: 62.0138
> Processing COMMENTS                CPU time: 28.5164
> Processing BOOKMARKS               CPU time: 0.0467
> Processing PROGRAM_ENTRY_POINTS    CPU time: 0.0006
> Processing SYMBOL_TABLE            CPU time: 12.4177
> Processing FUNCTIONS               CPU time: 15.8783
> Processing MARKUP                  
> ***** Exception occurred: XML Exporter failed! *****
>  TypeError: %X format: an integer is required, not NoneType
> ```
> 
> Unfortunately I'm not very experienced and more or less stumpled upon this, so if theres anything more I can do please feel free to tell me

I've updated idaxml.py again. Specifically, I added a None value check at the beginning of the write_numeric_attribute function. If it's convenient for you, please test it again. Thank you very much!

ryanmkurtz commented:
So we think all known issues have been addressed now?

Der-Systemfehler commented:
I had no more issues with exporting from my side

benbaker76 commented:
This commit along with that most recent fix works for me now too. Thanks @ZERO-A-ONE 🙏

ryanmkurtz commented:
Thanks everyone!
